### PR TITLE
ACCUMULO-4655 Added a Response Time column to the monitor which shows…

### PR DESCRIPTION
**Pull request by ivakegg**
… how long it took to get status from that tserver.  This is useful in determining whether long Last Contact times are dues to one or more tservers or whether this is due to churn in the master.

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java [hunk 1]
```diff
@@ -120,11 +120,17 @@ public TabletServerStatus getTableMap(boolean usePooledConnection) throws TExcep
       if (usePooledConnection == true)
         throw new UnsupportedOperationException();
 
+      long start = System.currentTimeMillis();
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java]: Using `System.nanoTime()` would be better here. More accurate and not subject to system clock change.

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java [hunk 2]
```diff
@@ -314,13 +314,22 @@ static void doTserverList(HttpServletRequest req, StringBuilder sb, List<TabletS
...
+    double avgResponseTime = 0.;
+    int count = 0;
     for (TabletServerStatus status : tservers) {
+      count++;
       avgLastContact += (now - status.lastContact);
+      avgResponseTime += status.responseTime;
     }
-    final long MINUTES = 3 * 60 * 1000;
+    avgResponseTime /= count;
+    final long SECOND = 1000;
```
Comment (joshelser) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: Make these static final at the class-level, please.

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java [hunk 3]
```diff
@@ -314,13 +314,22 @@ static void doTserverList(HttpServletRequest req, StringBuilder sb, List<TabletS
     long now = System.currentTimeMillis();
 
     double avgLastContact = 0.;
+    double avgResponseTime = 0.;
+    int count = 0;
     for (TabletServerStatus status : tservers) {
+      count++;
       avgLastContact += (now - status.lastContact);
+      avgResponseTime += status.responseTime;
```
Comment (joshelser) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: Did you give any thought to including other statistics than just avg? I am thinking about the case where there are one or a few outliers which significantly skew the mean. p95/p99 are nice to give some context, but admittedly require a little more work to compute.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 4]
```diff
@@ -2852,6 +2852,7 @@ public void run() {
   }
 
   public TabletServerStatus getStats(Map<String,MapCounter<ScanRunState>> scanCounts) {
+    long start = System.currentTimeMillis();
```
Comment (joshelser) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: Can use nanoTime here again.

File: server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java (hunk 1 above)
Comment (ivakegg) [inline server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java]: interesting...I will have to look that one up.  Thanks.

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java (hunk 3 above)
Comment (ivakegg) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: I was going with the mechanism already used for the lastContact time.  I will think on this.....

File: server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java (hunk 1 above)
Comment (ivakegg) [inline server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java]: Ok, it appears to be more accurate however it is in general slower depending on how the nanoTime is implemented (hardware dependent).  Not too sure we really need that time of accuracy here.....

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java (hunk 2 above)
Comment (ivakegg) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: right, of course.  Thanks.

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java (hunk 3 above)
Comment (ivakegg) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: What I cannot figure out here is that the avgLastContact is not actually an average...it is never divided by count (which I added).

File: server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java (hunk 1 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/master/LiveTServerSet.java]: I've started preferring to use nanoTime() for small, single-threaded timings. I know that's the sort of thing that it's meant to do well (local timings), but I'm not sure if there's a downside to it. Just thought I'd mention it. Like you say, at the end of the day, it probably won't matter.

I think @keith-turner had originally turned me onto it.

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java (hunk 3 above)
Comment (joshelser) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: Hahaha, that's a good catch :)

File: server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java [hunk 5]
```diff
@@ -309,18 +309,28 @@ public static void doDeadServerTable(HttpServletRequest req, StringBuilder sb, T
...
     }
-    final long MINUTES = 3 * 60 * 1000;
+    avgResponseTime /= count;
     tServerList.addSortableColumn("Server", new TServerLinkType(), null);
     tServerList.addSortableColumn("Hosted&nbsp;Tablets", new NumberType<>(0, Integer.MAX_VALUE), null);
-    tServerList.addSortableColumn("Last&nbsp;Contact", new DurationType(0l, (long) Math.min(avgLastContact * 4, MINUTES)), null);
+    tServerList.addSortableColumn("Last&nbsp;Contact", new DurationType(0l, (long) Math.min(avgLastContact * 4, LAST_CONTEXT_MAX_ERR)), null);
```
Comment (ivakegg) [inline server/monitor/src/main/java/org/apache/accumulo/monitor/servlets/TServersServlet.java]: ...and why are we multiplying the (not so) average by 4?

**Comment by milleruntime** [comment]
@ivakegg Are you planning on merging this forward to master?  We have yet to merge the new Monitor so I am wondering what would be the best technique to make sure this doesn't get lost.  Maybe open another PR for master and keep it open until the Monitor PR gets merged?

**Comment by ivakegg** [comment]
@milleruntime Yes, I would like this feature to persist.  I expect this would be merged into the master when complete.  If that is before the Monitor PR gets merge, then the person to merge the Monitor PR will have to resolve the conflicts.  Otherwise the person that merges this PR will have to resolve the conflicts.

**Comment by milleruntime** [comment]
Ok sounds good.  I was just thinking ahead since TServersServlet.java is removed in that PR.  It shouldn't be a problem finding a spot for your new stuff though... it should actually be easier with the new REST code.  A quick glance and it seems like here might be a good spot: https://github.com/apache/accumulo/pull/242/files#diff-a5f1cdfdb1d00ef6ab74e7e25cd35ce9

**Comment by ivakegg** [comment]
OK, I am going to attempt to merge this in tomorrow unless there are any more comments.

**Pull request closed without merging**